### PR TITLE
Chapter 9: Advanced login

### PR DESCRIPTION
**Pull request by ChuThom610**
chap9

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -3,8 +3,13 @@ def new; end
 
   def create
     user = User.find_by email: params[:session][:email].downcase
-    if user&.authenticate(params[:session][:password])
+    if user&.authenticate params[:session][:password]
       log_in user
+      if params[:session][:remember_me] == Settings.session.create.remember_me
+        remember(user)
```
Comment (thanhhungchu95) [inline app/controllers/sessions_controller.rb]: remove ()

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -3,8 +3,13 @@ def new; end
...
+      else
+        forget(user)
```
Comment (thanhhungchu95) [inline app/controllers/sessions_controller.rb]: remove ()

File: app/helpers/sessions_helper.rb [hunk 3]
```diff
@@ -3,15 +3,36 @@ def log_in user
...
 
+  def remember user
+    user.remember
+    cookies.permanent.signed[:user_id] = user.id
+    cookies.permanent[:remember_token] = user.remember_token
+  end
+
   def current_user
-    @current_user ||= User.find_by id: session[:user_id]
+    if (user_id = session[:user_id])
```
Comment (thanhhungchu95) [inline app/helpers/sessions_helper.rb]: remove ()

File: app/helpers/sessions_helper.rb [hunk 4]
```diff
@@ -3,15 +3,36 @@ def log_in user
...
+      @current_user ||= User.find_by id: user_id
+    elsif (user_id = cookies.signed[:user_id])
```
Comment (thanhhungchu95) [inline app/helpers/sessions_helper.rb]: remove ()

File: app/helpers/sessions_helper.rb [hunk 5]
```diff
@@ -3,15 +3,36 @@ def log_in user
...
   end
 
+  def forget user
+    user.forget
+    cookies.delete :user_id
+    cookies.delete :remember_token
+  end
+
   def log_out
+    forget(current_user)
```
Comment (thanhhungchu95) [inline app/helpers/sessions_helper.rb]: remove ()

File: app/models/user.rb [hunk 6]
```diff
@@ -12,6 +13,34 @@ class User < ApplicationRecord
...
                        length: {minimum: Settings.model.users.password.minimum}
 
+  def self.digest string
+    cost =  if ActiveModel::SecurePassword.min_cost
+              BCrypt::Engine::MIN_COST
+            else
+              BCrypt::Engine.cost
+            end
+
+    BCrypt::Password.create(string, cost: cost)
```
Comment (thanhhungchu95) [inline app/models/user.rb]: remove ()

File: app/models/user.rb [hunk 7]
```diff
@@ -12,6 +13,34 @@ class User < ApplicationRecord
...
+  end
+
+  def self.new_token
+    SecureRandom.urlsafe_base64
+  end
+
+  def remember
+    self.remember_token = User.new_token
+    update_attribute :remember_digest, User.digest(remember_token)
```
Comment (thanhhungchu95) [inline app/models/user.rb]: use update_attributes instead of this

File: app/models/user.rb [hunk 8]
```diff
@@ -12,6 +13,34 @@ class User < ApplicationRecord
...
+  end
+
+  def authenticated? remember_token
+    return false if remember_digest.nil?
+    BCrypt::Password.new(remember_digest).is_password?(remember_token)
+  end
+
+  def forget
+    update_attribute :remember_digest, nil
```
Comment (thanhhungchu95) [inline app/models/user.rb]: same here, use update_attributes

File: app/models/user.rb [hunk 9]
```diff
@@ -12,6 +13,34 @@ class User < ApplicationRecord
...
+  end
+
+  def remember
+    self.remember_token = User.new_token
+    update_attribute :remember_digest, User.digest(remember_token)
+  end
+
+  def authenticated? remember_token
+    return false if remember_digest.nil?
+    BCrypt::Password.new(remember_digest).is_password?(remember_token)
```
Comment (thanhhungchu95) [inline app/models/user.rb]: remove ()

File: app/helpers/sessions_helper.rb [hunk 10]
```diff
@@ -3,15 +3,36 @@ def log_in user
...
+    cookies.permanent[:remember_token] = user.remember_token
+  end
+
   def current_user
-    @current_user ||= User.find_by id: session[:user_id]
+    if user_id = session[:user_id]
+      @current_user ||= User.find_by id: user_id
+    elsif user_id = cookies.signed[:user_id]
+      user = User.find_by id: user_id
+      if user&.authenticated?(cookies[:remember_token])
```
Comment (thanhhungchu95) [inline app/helpers/sessions_helper.rb]: I think the parentheses at here is unnecessary

File: app/views/sessions/new.html.erb [hunk 11]
```diff
@@ -11,6 +11,11 @@
       <%= f.label :password %>
       <%= f.password_field :password, class: 'form-control' %>
```
Comment (thanhhungchu95) [inline app/views/sessions/new.html.erb]: use double-quote please

File: app/models/user.rb [hunk 12]
```diff
@@ -12,6 +13,34 @@ class User < ApplicationRecord
   validates :password, presence: true,
                        length: {minimum: Settings.model.users.password.minimum}
 
+  def self.digest string
```
Comment (thanhhungchu95) [inline app/models/user.rb]: use syntax `class << self` instead

File: app/assets/stylesheets/custom.scss [hunk 13]
```diff
@@ -176,3 +176,17 @@ input {
...
+.checkbox {
+  margin-top: -10px;
+  margin-bottom: 10px;
+  span {
+    margin-left: 20px;
+    font-weight: normal;
+  }
+}
+
+#session_remember_me {
```
Comment (thanhhungchu95) [inline app/assets/stylesheets/custom.scss]: can you change this to `class` ? I don't think `id` is the good choice.

Comment (ChuThom610) [reply app/assets/stylesheets/custom.scss]: Can you explain to me why id is a good choice?

Comment (thanhhungchu95) [reply app/assets/stylesheets/custom.scss]: `class` is better, because It can be shared among many elements, while `id` can only be used for one element

Comment (ChuThom610) [reply app/assets/stylesheets/custom.scss]: Ohh, I'm sorry, I misread your comment. I understand, thank you so much.

**Pull request merged**